streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -294,11 +306,13 @@ public synchronized void start() {
         log.debug("Starting Kafka Stream process");
 
         if (state == KafkaStreams.State.CREATED) {
-            for (final StreamThread thread : threads)
-                thread.start();
-
             setState(KafkaStreams.State.RUNNING);
 
+            for (final StreamThread thread : threads) {
+                threadState.put(thread.getId(), thread.state());
```

Comment:
This should be `synchronized` too as you could have other started threads updating the map at the same time. Perhaps just wrap the map with `Collections.synchronizedMap(...)` and then remove `synchronized` from `StreamStateListener.onChange(..)`

Sorry for the back-and-forth.